File: mobile-ios/App/Sources/App/Session/NotificationsStore.swift
import Domain
import Networking
import Observation

/// Bildirimlerin uygulama genelindeki durumu (zil rozeti + bildirim ekranı).
@MainActor
@Observable
final class NotificationsStore {
    private(set) var items: [AppNotification] = []
    private(set) var isLoading = false

    @ObservationIgnored private let repository: any NotificationsRepository
    @ObservationIgnored private let session: SessionStore

    nonisolated init(repository: any NotificationsRepository, session: SessionStore) {
        self.repository = repository
        self.session = session
    }

    var unreadCount: Int {
        items.unreadCount
    }

    func refresh() async {
        guard session.isSignedIn else {
            items = []
            return
        }
        isLoading = true
        defer { isLoading = false }
        items = await (try? repository.notifications()) ?? items
    }

    func markRead(id: String) async {
        guard let index = items.firstIndex(where: { $0.id == id }),
              !items[index].isRead else { return }
        items[index] = markingRead(items[index])
        try? await repository.markRead(id: id)
    }

    func markAllRead() async {
        items = items.map(markingRead)
        try? await repository.markAllRead()
    }

    func clearLocal() {
        items = []
    }

    private func markingRead(_ notification: AppNotification) -> AppNotification {
        AppNotification(
            id: notification.id,
            type: notification.type,
            title: notification.title,
            body: notification.body,
            isRead: true,
            relatedOrderID: notification.relatedOrderID,
            createdAt: notification.createdAt
        )
    }
}

File: mobile-android/app/src/main/java/com/erenlermarket/app/data/session/NotificationsStore.kt
package com.erenlermarket.app.data.session

import com.erenlermarket.app.di.AppScope
import com.erenlermarket.app.domain.model.AppNotification
import com.erenlermarket.app.domain.model.unreadCount
import com.erenlermarket.app.domain.repository.NotificationsRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/** Bildirim durumu — zil rozeti için. Oturum kapanınca temizlenir. */
@Singleton
class NotificationsStore @Inject constructor(
    private val repository: NotificationsRepository,
    session: SessionManager,
    @AppScope private val scope: CoroutineScope,
) {

    private val _items = MutableStateFlow<List<AppNotification>>(emptyList())
    val items: StateFlow<List<AppNotification>> = _items.asStateFlow()

    val unreadCount: StateFlow<Int> = _items
        .map { it.unreadCount }
        .stateIn(scope, SharingStarted.Eagerly, 0)

    init {
        scope.launch {
            session.state.collect { state ->
                if (state is SessionState.SignedIn) refresh() else _items.value = emptyList()
            }
        }
    }

    suspend fun refresh() {
        runCatching { repository.notifications() }.onSuccess { _items.value = it }
    }

    suspend fun markRead(id: String) {
        _items.value = _items.value.map { if (it.id == id) it.copy(isRead = true) else it }
        runCatching { repository.markRead(id) }
    }

    suspend fun markAllRead() {
        _items.value = _items.value.map { it.copy(isRead = true) }
        runCatching { repository.markAllRead() }
    }
}
